mlir/lib/Dialect/Vector/Transforms/VectorDropLeadUnitDim.cpp

```diff
@@ -399,13 +399,28 @@ mlir::vector::castAwayContractionLeadingOneDim(vector::ContractionOp contractOp,
           transposeResults.push_back(targetExpr);
         }
       }
+
+      // Check if the transpose effects outer unit dims only. Such transposes do
+      // not materially effect the underlying vector and can be omitted.
+      bool tranposeNonOuterUnitDims = false;
+      for (int64_t i = 0; i < (int64_t)perm.size(); ++i) {
+        if (perm[i] != i && i != (int64_t)perm.size() - 1) {
+          if (operands[it.index()].getType().cast<ShapedType>().getDimSize(i) !=
+              1) {
```

Comment:
I think we can use `llvm::enumerate` and we can get rid of c-style casting. Also, we can collapse two `if` into a single if, so we avoid additional nesting.

```cpp
for (auto [idx, dim] : llvm::enumerate(perm)) {
  // ...
}
```